lib/ocpp/v201/smart_charging.cpp

```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
+                                                             const ocpp::DateTime& end_time, const int32_t evse_id,
+                                                             ChargingRateUnitEnum charging_rate_unit) {
+    CompositeSchedule composite_schedule;
+
+    composite_schedule.evseId = evse_id;
+    composite_schedule.duration = duration_cast<seconds>(end_time.to_time_point() - start_time.to_time_point()).count();
+    composite_schedule.scheduleStart = start_time;
+    composite_schedule.chargingRateUnit = charging_rate_unit;
+
+    return composite_schedule;
+}
+
+int32_t SmartChargingHandler::determine_duration(const ocpp::DateTime& start_time, const ocpp::DateTime& end_time) {
+    return duration_cast<seconds>(end_time.to_time_point() - start_time.to_time_point()).count();
+}
+
+bool SmartChargingHandler::within_time_window(const ocpp::DateTime& end_time, const ocpp::DateTime& temp_time) {
+    return SmartChargingHandler::determine_duration(temp_time, end_time) > 0;
+    // auto end_time_point = end_time.to_time_point();
+    // auto temp_time_point = temp_time.to_time_point();
+    // auto count = duration_cast<seconds>(end_time_point - temp_time_point).count();
+
+    // EVLOG_info << "";
+    // EVLOG_info << "time_window> " << temp_time.to_rfc3339() << " - " << end_time.to_rfc3339() << " = " << count << "
+    // ("
+    //            << get_log_duration_string(count) << ")";
+    // if (count > 0) {
+    //     return true;
+    // } else {
+    //     EVLOG_info << "time_window> count = " << count << " RETURN FALSE ";
+    //     return false;
+    // }
+}
+
+PeriodDateTimePair SmartChargingHandler::find_period_at(const ocpp::DateTime& time, const ChargingProfile& profile,
+                                                        const int evse_id) {
+    auto period_start_time = this->get_profile_start_time(profile, time, evse_id);
+
+    EVLOG_info << "#" << profile.id << " find_period_at> " << period_start_time.value().to_rfc3339();
+
+    // TODO: Only processing first charging schedule
+    const auto schedule = profile.chargingSchedule.front();
+
+    if (period_start_time) {
+        const auto periods = schedule.chargingSchedulePeriod;
+        time_point<date::utc_clock> period_end_time;
+        for (size_t i = 0; i < periods.size(); i++) {
+            const auto period_end_time = get_period_end_time(i, period_start_time.value(), schedule);
+
+            EVLOG_info << "   find_period_at>        start_time> " << time.to_rfc3339();
+            EVLOG_info << "   find_period_at> period_start_time> " << period_start_time.value().to_rfc3339();
+            EVLOG_info << "   find_period_at>   period_end_time> " << period_end_time.to_rfc3339();
+
+            if (time >= period_start_time.value() && time < period_end_time) {
+                PeriodDateTimePair date_time_pair = {periods.at(i), ocpp::DateTime(period_end_time)};
```

Comment:
Can you update this with the struct fields:
` {.period = periods.at(i) .end_time = ocpp::DateTime(period_end_time)};`